feat(curriculum): add equation solver project to SCWP

Dario-DC opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #XXXXX

<!-- Feel free to add any additional description of changes below this line -->

Dario-DC committed:
chore(curriculum): put Python certification projects in separate folders (#54004)

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

majestic-owl448 committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

Dario-DC committed:
create project

Dario-DC committed:
add first steps

Dario-DC committed:
add steps 8-11

zairahira committed:
feat(curriculum): add bisection method project to Python (#54110)

Co-authored-by: Dario-DC <105294544+Dario-DC@users.noreply.github.com>
Co-authored-by: Dario-DC <dicillodario@gmail.com>

Dario-DC committed:
add steps 8-15

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into reorder-python

Dario-DC committed:
merge reorder-python

Dario-DC committed:
add __str__ implementation

Dario-DC committed:
fix instructions + add tests step 1-19

Dario-DC committed:
fix instructions

Dario-DC committed:
merge main

solve conflicts

Dario-DC committed:
fix steps 1-26

Dario-DC committed:
fix steps 27 -29

Dario-DC committed:
add steps 30-32

Dario-DC committed:
add instructions step 33-37

Dario-DC committed:
add description step 32

Dario-DC committed:
fix tests until step 40

Dario-DC committed:
add instructions until step 45

Dario-DC committed:
add tests until step 49

Dario-DC committed:
add step 50

Dario-DC committed:
fix instructions and tests until step 50

Dario-DC committed:
add final steps

Dario-DC committed:
add missing tests/hints

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into equation-solver

Dario-DC committed:
remove isBeta

Dario-DC committed:
fix instructions

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into equation-solver

Dario-DC committed:
delete time, template, required from meta.json

zairahira committed:
fix: nits

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662fab95a3f903a7aa028550.md:
````diff
@@ -0,0 +1,71 @@
+trout.swim # Swimming!
+```
+
+In the example above, the `swim` property of `trout` is accessed throught `trout.swim` instead of `trout.swim()`, as it would be for a regular method.
````
```suggestion
In the example above, the `swim` property of `trout` is accessed through `trout.swim` instead of `trout.swim()`, as it would be for a regular method.
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662fbcef5f05e1b84f541a0c.md:
```diff
@@ -0,0 +1,75 @@
+# --description--
+
+Each equation object will be instantiated passing as many arguments as the coefficients of the equation, starting from n-th degree of \\( x \\) down to the zero-th degree, including the possible coefficient with the value of `0`.
+
```
Maybe we can add something like this to further show what the instruction means:

```md
For example, (4,5) would translate to \\( 4x^1 + 5x^0 = 0 \\) or  \\( 4x + 5 = 0 \\) with `4` being the coefficient of the first(highest here) degree and `5` the coefficient of the zero-th degree.
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662a6bc12cde72c32fb526f0.md:
```diff
@@ -0,0 +1,30 @@
+
+# --description--
+
+An *interface* is like a blueprint for a class. An interface contains a set of methods and properties that a class should implement.  
```
I came up with this intro:

```suggestion
In programming, an *interface* is like a blueprint for a class that contains a set of methods and properties that a class should implement. Interfaces facilitate communication between different parts of a program without detailing their internal workings, hence promoting encapsulation and code re-usability.
```

zairahira replied:
```suggestion
An <dfn>interface</dfn> is like a blueprint for a class. An interface contains a set of methods and properties that a class should implement.  
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662bd456896f16d9bd03f1a6.md:
```diff
@@ -0,0 +1,47 @@
+
+# --description--
+
+Within the `Equation` class, define two new instance methods named `solve` and `analyze`.
```
Have we stopped testing if the user has removed the `pass` keyword?
Noticed the same in step-10.

Dario-DC replied:
For the new additions to the curriculum we are not testing it. Probably, after the first projects, we should stop asking to add `pass` and then replace `pass` with `<x>`... It breaks the flow and complicate the instructions. It has no effect, so once campers understand why we are using it, I think there's no point in asking them to remove it.

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662bdd364bf2cde1487922a9.md:
```diff
@@ -0,0 +1,43 @@
+
+Unlike other programming languages, Python does not implement interfaces in its core language, but Python standard library allows you to define interfaces in a simple way.
+
+For this project, you'll use utilities from the `abc` module. So, go at the top of your code and import `ABC` from the `abc` module.
```
It is better to mention what `ABC` is in this step rather than the next step.

zairahira replied:
Also, the editable region is already at the top so the user shouldn't be instructed to go to the top.

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662bdd364bf2cde1487922a9.md:
```diff
@@ -0,0 +1,43 @@
+
+--fcc-editable-region--
+class Equation:
+    def solve(self): pass
+        
+    def analyze(self): pass
```
Is this code orientation intentional?

Dario-DC replied:
Yeah, but I'm not mentioning it can be done so I'm thinking of restoring the usual indentation.

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662bde88dc84f1e249801b1a.md:
```diff
@@ -0,0 +1,41 @@
+
+# --description--
+
+`ABC` stands for *Abstract Base Classes*. The `ABC` class enables you to turn a regular class into an abstract class that acts as a blueprint for concrete classes.
```
Have we mentioned about concrete classes and abstraction before?

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663b83a28943e6aa6275a514.md:
```diff
@@ -0,0 +1,93 @@
+
+Still within the `Equation` class, define a `__str__` method to give a proper string representation to the equation objects you are going to create.
+
+For now, declare a variable `terms` and assign it an empty list. You'll use this variable to store each term (coefficient times \\( x^n \\)) of your equation.
```
```suggestion
For now, within the `__str__` method, declare a variable `terms` and assign it an empty list. You'll use this variable to store each term (coefficient times \\( x^n \\)) of your equation.
```

zairahira reviewed (commented):
Reviewed till step 20.

zairahira committed:
fix: inline code in step 11

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/664e4a590b52ba8d2adff19f.md:
```diff
@@ -0,0 +1,129 @@
+
+The discriminant of a quadratic equation in the form \\( ax^2 + bx + c = 0 \\), usually indicated by the capital Greek letter delta, is equal to \\( Δ = b^2 - 4ac \\).
+
+Within the `QuadraticEquation` class, define an `__init__` method. Use a `super()` call to inherit the `__init__` method defined inside the interface and define a new attribute named `delta`, which stores the value of the discriminant of the equation.
```
Note: `super()` is introduced in the vector space, and this would determine the order of this project.

Dario-DC replied:
This should come just after the vector space :+1:

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/664eef158d792a509e8d708a.md:
```diff
@@ -0,0 +1,115 @@
+
+It takes three arguments: the regex pattern to match, the replacement, and the string on which you want to perform the replacement.
+
+From your `__str__` function, return a `sub()` call passing the string `'1'`, an empty string, and your existing `equation_string.strip('+')` call as the arguments. This will replace each `1` with an empty string. Don't worry, as you'll refine the regex pattern in the next steps.
```
```suggestion
From your `__str__` function, return a `sub()` call passing the string `'1'`, an empty string, and your existing `equation_string.strip('+')` call as the arguments. This will replace each `1` with an empty string. The result is not refined yet and you'll continue to work on the regex pattern in the next steps.
```

zairahira reviewed (commented):
Reviewed till step 40.

Dario-DC commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663a32735b317af9812eb0d7.md:
```diff
@@ -0,0 +1,117 @@
+
+In linear equations in the form \\( ax + b = 0 \\), the slope is simply the coefficient \\( a \\), and the y-intercept is the coefficient \\( b \\).
+
+You are going to use the `analyze` method to provide additional information about the equation. Inside the `analyze` method, append a string in the form `f'slope = {}\ny-intercept = {}'` to the `details` attribute: fill the curly braces so that they evaluate to the slope and the y-intercept, respectively.
```
It's expected, since we are printing the list. If we print the item it shows in separate lines. Maybe I can ask to print the string instead of the list.

Dario-DC commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663c9f31306353460da54542.md:
```diff
@@ -0,0 +1,82 @@
+
+# --description--
+
+As you can see, the `+` sign is missing from the output. The number sign is displayed by default only if negative. To change this behaviour, you can write a colon after the expression to be evaluated within the curly braces of your f-string, and specify the option `+`. This will allow you to display the sign both for positive and negative numbers.
```
The colon is not part of the format specifier.

Dario-DC committed:
implemented suggestions from review

zairahira reviewed (commented):
Reviewed all the steps. 

Some general suggestions:

It should be mentioned somewhere that the project covers equations up to second degree. A good place would be to add in intro.

zairahira commented:
Reminders:
- Add intro. 
- Flesh out pages.
- Verify project order.

gikf commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/6656262ec602ff688c50c8a9.md:
```diff
@@ -0,0 +1,165 @@
+
+# --description--
+
+After the `if` statement you wrote in the previous step, ensure that `solve()` and `analyze()` methods have not been called yet, otherwise you would keep adding the same items to the `results` and `details` lists.
+
+Check if the `results` and `details` attributes of `equation` contain elements. If not, call the `solve` method, and the `analyze` method, respectively.
```
Using class this way is a bit more complicated than it should. Couldn't `solver` function get the _results_ from what `solve` method would return, and _details_ from what `analyze` would return? The strange relation between two is also visible later, when `analyze` is changed to save strings aligned to specifically match `solver`'s formatting.

Both these methods could keep the calculations internally saved, so it would be done just once in any case. Ie. when `solve` would be called again, it would just return the saved result.

gikf commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/6656e5541f2db163312505d6.md:
```diff
@@ -0,0 +1,165 @@
+        equation.analyze()
+--fcc-editable-region--
+    output_string = f'\n{equation.type:-^24}'
+    output_string += f'\n\n{str(equation):^24}\n\n'
```
```suggestion
    output_string += f'\n\n{equation!s:^24}\n\n'
```
With various f-string formatting _tricks_ `!s` could also be thrown here.

Dario-DC commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/662fa2e2cf27c09f21f4f5d0.md:
```diff
@@ -0,0 +1,68 @@
+
+# --description--
+
+Inside the `Equation` class, define a new abstract method `degree`. Then, define it within the `LinearEquation` class to avoid errors.
```
Assuming we change both `type` and `degree` into class attributes, could something like the following work?
```py
class Equation(ABC):
    type: str
    degree: int
    
   def __init__(self. *args):
      #current __init__ body   

   def __init_subclass__(cls):
      if not (hasattr(cls, "type") and hasattr(cls, "degree")):
        raise AttributeError("<error message>")
```

Dario-DC commented:
Based on the feedback received, I modified the [prototype](https://github.com/freeCodeCamp/CurriculumExpansion/pull/372) . Here are the main changes:
- replaced abstract properties `degree` and `type` with class attributes
- added `__init_subclass__` method (not sure about the implementation)
- modified `coefficients` attribute to keep the same order of `args`
- removed `self.coefficients[<n>]` references, (coefficients are now unpacked to improve readability)
- moved the creation of the strings to print outside the `solve` and `analyze` methods.

I'll start to implement these changes here as soon as I get a positive feedback. @ilenia-magoni @zairahira @gikf

Dario-DC committed:
resolve conflict

Dario-DC committed:
fix steps up to 25

Dario-DC committed:
fix project up to step 33

Dario-DC committed:
fix project up to step 43

Dario-DC committed:
fix project up to step 48

Dario-DC commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/664ef4623946e65e18d59764.md:
```diff
@@ -0,0 +1,125 @@
+
+# --description--
+
+In a regex pattern, a *lookaround* is an assertion that matches a certain pattern without consuming characters in the string. One kind of lookaround is the lookbehind, which can be either positive or negative. They are denoted by `(?<=...)` and `(?<!...)`, respectively.
```
The idea was to teach lookarounds.

Dario-DC commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/6650f037c017aa6855a608e3.md:
```diff
@@ -0,0 +1,123 @@
+
+# --hints--
+
+You should return a dictionary containing two keys, `'x'`, and `'y'`, and the corresponding values of vertex \\( x \\) and \\( y \\) coordinates, respectively.
```
It seems it's not rendered in hints :thinking:

Dario-DC committed:
fix remaining steps

[skip ci]

Dario-DC committed:
add missing tests

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into equation-solver

Dario-DC committed:
add intro + new step 15

naomi-lgbt commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663b7fefd437bd984e091cbf.md:
```diff
@@ -1,13 +1,13 @@
 # --description--
 
-Next, create a new class named `QuadraticEquation` and make it inherit from `Equation`. You'll use this new class to represent quadratic equations.
+Next, create a new class named `QuadraticEquation` and make it inherit from `Equation`. You'll use this new class to represent quadratic equations, which are second-degree equation having the form $ax^2 + bx + c = 0$. 
```
```suggestion
Next, create a new class named `QuadraticEquation` and make it inherit from `Equation`. You'll use this new class to represent quadratic equations, which are second-degree equations having the form $ax^2 + bx + c = 0$. 
```

naomi-lgbt reviewed (commented):
Grammar nit

Dario-DC committed:
Update curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663b7fefd437bd984e091cbf.md

Co-authored-by: Naomi the Technomancer <accounts+github@nhcarrigan.com>

Dario-DC committed:
add space in editable region

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/663a32735b317af9812eb0d7.md:
```diff
@@ -0,0 +1,106 @@
+
+<img alt="a plot of a linear function" src="https://cdn.freecodecamp.org/curriculum/python/linear-equation.png" style="background-color: white; height: 350px; width: auto; padding: 15px; display: block; margin-right: auto; margin-left: auto; margin-bottom: 1.2rem;">
+
+You are going to use the `analyze` method to provide additional information about the equation. Inside the `analyze` method, unpack the coefficeints into the variables `slope` and `intercept`.
```
```suggestion
You are going to use the `analyze` method to provide additional information about the equation. Inside the `analyze` method, unpack the coefficeints into the variables `slope` and `intercept`.
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/664e4e1b6c35a99cbba49e84.md:
```diff
@@ -0,0 +1,105 @@
+
+Now, create an instance of the `QuadraticEquation` class to represent the equation \\( 11x^2 - x + 1 = 0 \\).
+
+Assign the new instance to a variable `quadr_eq`, then print your new variable.
```
```suggestion
Assign the new instance to a variable `quadr_eq`, then print your new variable. Note that, at this point, the second degree term would be missing from the string representation of the equation.
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/66799c1a0204668cef35555d.md:
```diff
@@ -0,0 +1,157 @@
+
+# --description--
+
+f-strings also enable you to set a specific precision to your numerical data by using the `.nf` the format specifier, where `n` is the number of decimal digits to display.
```
```suggestion
f-strings also enable you to set a specific precision to your numerical data by using the `.nf` format specifier, where `n` is the number of decimal digits to display.
```

zairahira commented on curriculum/challenges/english/07-scientific-computing-with-python/learn-interfaces-by-building-an-equation-solver/667944fed1f6b61da3406bd8.md:
```diff
@@ -0,0 +1,164 @@
+
+# --description--
+
+The structural pattern matching is a Python construct that enables to match a pattern with a subject value, which is specified after the `match` keyword:
```
```suggestion
Structural pattern matching is a Python construct that enables matching a pattern with a subject value, which is specified after the `match` keyword:
```

zairahira reviewed (approved):
I found some nits.
The project is looking great overall, let's get this in 🎊

zairahira commented on client/i18n/locales/english/intro.json:
```diff
@@ -884,6 +884,14 @@
+      "learn-interfaces-by-building-an-equation-solver": {
+        "title": "Learn Interfaces by Building an Equation Solver",
+        "intro": [
+          "Abstract classes enables you to define formal interfaces - objects acting as blueprints for classes.",
```
```suggestion
          "Abstract classes enable you to define formal interfaces - objects acting as blueprints for classes.",
```

Dario-DC committed:
Merge branch 'equation-solver' of https://github.com/Dario-DC/freeCodeCamp into equation-solver

Dario-DC committed:
Apply suggestions from code review

Co-authored-by: Zaira <33151350+zairahira@users.noreply.github.com>

Dario-DC committed:
Merge branch 'equation-solver' of https://github.com/Dario-DC/freeCodeCamp into equation-solver

Dario-DC committed:
fix typo

Dario-DC committed:
make projectile trajectory project visible

Dario-DC committed:
Merge branch 'main' of https://github.com/freeCodeCamp/freeCodeCamp into equation-solver

Dario-DC commented on curriculum/challenges/_meta/learn-interfaces-by-building-an-equation-solver/meta.json:
```diff
@@ -0,0 +1,280 @@
+{
+  "name": "Learn Interfaces by Building an Equation Solver",
+  "isUpcomingChange": true,
```
```suggestion
  "isUpcomingChange": false,
```

Dario-DC committed:
Update curriculum/challenges/_meta/learn-interfaces-by-building-an-equation-solver/meta.json
